lesson_4/lesson_4.py

```diff
@@ -0,0 +1,166 @@
+"""
+Написать программу, которая скачивает изображения с заданных URL-адресов и сохраняет их на диск.
+Каждое изображение должно сохраняться в отдельном файле, название которого
+соответствует названию изображения в URL-адресе.
+
+Например, URL-адрес: https://example/images/image1.jpg -> файл на диске: image1.jpg
+
+— Программа должна использовать многопоточный, многопроцессорный и асинхронный подходы.
+— Программа должна иметь возможность задавать список URL-адресов через аргументы командной строки.
+— Программа должна выводить в консоль информацию о времени скачивания каждого изображения и
+ общем времени выполнения программы.
+"""
+
+import argparse
+import asyncio
+import aiohttp
+import requests
+import threading
+from urllib.parse import urlparse
+from multiprocessing import Process
+
+import os
+import time
+
+
+# Default URLs
+exp_urls = [
+        'https://mykaleidoscope.ru/x/uploads/posts/2022-10/'
+        '1666364979_14-mykaleidoscope-ru-p-krasivie-peizazhi-prirodi-oboi-17.jpg',
+        'https://mykaleidoscope.ru/x/uploads/posts/2022-10/'
+        '1666365039_2-mykaleidoscope-ru-p-krasivie-peizazhi-prirodi-oboi-4.jpg',
+        'https://gas-kvas.com/uploads/posts/2023-02/'
+        '1675483689_gas-kvas-com-p-fonovii-risunok-dlya-kompyutera-priroda-15.jpg',
+        'https://catherineasquithgallery.com/uploads/posts/2021-03/'
+        '1614612233_137-p-fon-dlya-fotoshopa-priroda-209.jpg',
+        ]
+
+
+def download_image(url, folder, start_time):
+    """uploading photos to a folder"""
+    try:
+        response = requests.get(url, stream=True)
+        if response.status_code == 200:
+            # Extracting the file name from the URL
+            parsed_url = urlparse(url)
+            filename = os.path.join(folder, os.path.basename(parsed_url.path))
+
+            # Saving the image
+            with open(filename, 'wb') as file:
+                for chunk in response.iter_content(chunk_size=1024):
+                    file.write(chunk)
```

Comment:
Как то у вас долго выполняется код. Скорее всего из-за этого
```suggestion
        response = requests.get(url)
        if response.status_code == 200:
            # Extracting the file name from the URL
            filename = os.path.join(folder, os.path.basename(url))

            # Saving the image
            with open(filename, 'wb') as file:
                file.write(response.content)
```
После этого у меня скорость возросла в 6 раз, в основном это сказалось на многопоточном подходе